### ISSUE TITLE
Incorrect result in ghci or when compiled with -O0

### ISSUE THREAD

**Issue by nh2** [swamp-agr/blaze-textual#11]
```
> import Blaze.ByteString.Builder (toByteString)
> import Blaze.Text (integral)
> import Data.Int

> toByteString (integral (-127 :: Int8))
"-127"

> toByteString (integral (-128 :: Int8))
"-\176"
```

It should print `-128`, but does so only on `-O1` or higher.

Note that `-128` is a valid `Int8`, there is no overflow involved.

**Comment by nh2** [comment]
This is because this code is incorrect:

https://github.com/bos/blaze-textual/blob/9748b60d7ecfb2009eb51f68fe79e6bec211c574/Blaze/Text/Int.hs#L50-L58

It computes `nonNegative (-i)`.

But:

```
> - (-128 :: Int8)
-128
```

This means that `nonNegative` is called on a function that's negative, so that function breaks.

**Comment by nh2** [comment]
The code for `nonNegative` is: https://github.com/bos/blaze-textual/blob/9748b60d7ecfb2009eb51f68fe79e6bec211c574/Blaze/Text/Int.hs#L86-L89

This bug could have been avoided by inserting a check for the precondition that the number is not negative, for example:

```haskell
nonNegative x
  | x < 0 = error $ "nonNegative: Called with negative number " ++ show (fromIntegral x :: Integer)
  | otherwise = go x
  where
    go n | n < 10    = digit n
         | otherwise = go (n `quot` 10) `mappend` digit (n `rem` 10)
```

It could also have been found by running the test suite once in ghci, which would probably have been a good idea for a code base that makes heavy use of rewrite rules and specifically acknowledges that the code path for ghci is different than for compiled code.

**Comment by nh2** [comment]
The comment in the code in https://github.com/bos/blaze-textual/issues/11#issuecomment-493722651 says

> The test for "-0" catches an overflow if we render minBound.

It was added in commit e43432c495d3ed2211310615457ed9788d5bc09d - _"Fix the rendering of minBound on bounded integral types"_, which also added the code

```haskell
    | toByteString b == "-0" = fromString (show i)
```

But I don't understand it, how was that supposed to work? How should `toByteString b` return `"-0"`?

Was `nonNegative` perhaps supposed to be implemented like this?

```haskell
nonNegative x
  | x < 0 = "0"
  | otherwise = go x
  where
    go n | n < 10    = digit n
         | otherwise = go (n `quot` 10) `mappend` digit (n `rem` 10)
```

but that extra case was forgotten?

**Comment by nh2** [comment]
PR to fix it in #12.
